File: JavaExpenseApp/src/DbConnection.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

public class DbConnection {
    private static final String URL = "jdbc:mysql://localhost:3306/test";
    private static final String USER = "root";
    private static final String PASSWORD = "";
    private static Connection connection = null;

    // Establish a connection to the database
    public static Connection connect() {
        if (connection == null) {
            try {
                connection = DriverManager.getConnection(URL, USER, PASSWORD);
                System.out.println("Database connection established.");
            } catch (SQLException e) {
                System.err.println("Failed to connect to the database: " + e.getMessage());
            }
        }
        return connection;
    }

    // Close the database connection
    public static void disconnect() {
        if (connection != null) {
            try {
                connection.close();
                connection = null;
                System.out.println("Database connection closed.");
            } catch (SQLException e) {
                System.err.println("Failed to close the database connection: " + e.getMessage());
            }
        }
    }
}

File: db_connection.py
import mysql.connector  # Importa il modulo per la connessione a MySQL

class DbConnection:
    _connection = None  # Variabile di classe per mantenere la connessione aperta (singleton)

    @staticmethod
    def connect():
        # Metodo statico per ottenere la connessione al database
        if DbConnection._connection is None:  # Se non esiste già una connessione
            try:
                DbConnection._connection = mysql.connector.connect(
                    host="localhost",        # Indirizzo del server MySQL
                    user="root",             # Nome utente del database
                    password="ccgZ86+d$L:i*Ez",  # Password del database
                    database="project"       # Nome del database da usare
                )
                print("Database connection established.")  # Messaggio di conferma
            except mysql.connector.Error as e:
                print(f"Failed to connect to database: {e}")  # Messaggio di errore in caso di problemi
                DbConnection._connection = None  # Imposta la connessione a None in caso di errore
        return DbConnection._connection  # Restituisce la connessione (nuova o già esistente)

    @staticmethod
    def disconnect():
        # Metodo statico per chiudere la connessione al database
        if DbConnection._connection is not None:  # Se esiste una connessione aperta
            try:
                DbConnection._connection.close()  # Chiude la connessione
                DbConnection._connection = None   # Reset della variabile di classe
                print("Database connection closed.")  # Messaggio di conferma
            except mysql.connector.Error as e:
                print(f"Failed to close the database connection: {e}")  # Messaggio di errore in caso di problemi